From 48c555edb87cdf8bc33a6d47c83ea79ef7cfa87d Mon Sep 17 00:00:00 2001
From: Wenyou Yang <wenyou.yang@atmel.com>
Date: Wed, 28 Jan 2015 14:05:32 +0800
Subject: [PATCH] driver: at91_twi: add configure the TWI bus by the board
 support.

Let the board determines which TWI bus that the HDMI,
Audio Codec, PMIC ACT8865 and EEPROM is wired on, instead of through menuconfig,
in order to reduce the menuconfig complexity.

The function at91_board_config_twi_bus() is provided by the board file

Signed-off-by: Wenyou Yang <wenyou.yang@atmel.com>
---
 Config.in                                 | 21 +++++++++++++++++----
 board/sama5d3_xplained/sama5d3_xplained.c |  8 ++++++++
 board/sama5d4_xplained/sama5d4_xplained.c | 11 +++++++++++
 board/sama5d4ek/sama5d4ek.c               | 10 ++++++++++
 driver/act8865.c                          | 12 ++++++++----
 driver/at24xx.c                           | 12 ++++++++----
 driver/at91_twi.c                         | 16 ++++++++++++++++
 driver/driver_cpp.mk                      |  3 +++
 driver/hdmi_SiI9022.c                     | 12 ++++++++----
 driver/wm8904.c                           | 12 ++++++++----
 include/board.h                           |  1 +
 include/twi.h                             |  5 +++++
 12 files changed, 103 insertions(+), 20 deletions(-)

diff --git a/Config.in b/Config.in
index 5780124..7864057 100644
--- a/Config.in
+++ b/Config.in
@@ -408,7 +408,7 @@ config CONFIG_HDMI
 
 choice
 	prompt "HDMI on TWI Select"
-	depends on CONFIG_HDMI
+	depends on CONFIG_HDMI && !CONFIG_AUTOCONFIG_TWI_BUS
 	default CONFIG_HDMI_ON_TWI0
 	help
 	  This interface allow you select which TWI bus for the HDMI to wire on.
@@ -446,7 +446,7 @@ config CONFIG_WM8904
 
 choice
 	prompt "CODEC on TWI Select"
-	depends on CONFIG_WM8904
+	depends on CONFIG_WM8904 && !CONFIG_AUTOCONFIG_TWI_BUS
 	default CONFIG_CODEC_ON_TWI0
 	help
 	  This interface allow you select which TWI bus for the CODEC to wire on.
@@ -507,7 +507,7 @@ config CONFIG_ACT8865
 
 choice
 	prompt "PMIC (ACT8865) on TWI Select"
-	depends on CONFIG_ACT8865
+	depends on CONFIG_ACT8865 && !CONFIG_AUTOCONFIG_TWI_BUS
 	default CONFIG_PMIC_ON_TWI0
 	help
 	  This interface allow you select which TWI bus for the PMIC to wire on.
@@ -556,7 +556,7 @@ config CONFIG_LOAD_EEPROM
 
 choice
 	prompt "EEPROM on TWI Select"
-	depends on CONFIG_LOAD_EEPROM
+	depends on CONFIG_LOAD_EEPROM && !CONFIG_AUTOCONFIG_TWI_BUS
 	default CONFIG_EEPROM_ON_TWI0
 	help
 	  This interface allow you select which TWI bus for the EEPROM to wire on.
@@ -584,3 +584,16 @@ config	CONFIG_EEPROM_ON_TWI3
 endchoice
 
 endmenu
+
+config CONFIG_AUTOCONFIG_TWI_BUS
+	bool "Auto Configure TWI Bus by Board"
+	depends on CONFIG_TWI
+	select CONFIG_TWI0 if CPU_HAS_TWI0
+	select CONFIG_TWI1 if CPU_HAS_TWI1
+	select CONFIG_TWI2 if CPU_HAS_TWI2
+	select CONFIG_TWI3 if CPU_HAS_TWI3
+	default y
+	help
+	  This interface allow the board determines  which TWI bus that
+	  the HDMI, Audio CODEC(wm8904), PMIC(ACT8865) is wired on,
+	  select automatically, no menuconfig selecting to need.
diff --git a/board/sama5d3_xplained/sama5d3_xplained.c b/board/sama5d3_xplained/sama5d3_xplained.c
index fbaf35e..b36b6aa 100644
--- a/board/sama5d3_xplained/sama5d3_xplained.c
+++ b/board/sama5d3_xplained/sama5d3_xplained.c
@@ -45,6 +45,7 @@
 #include "arch/at91_ddrsdrc.h"
 #include "sama5d3_xplained.h"
 #include "act8865.h"
+#include "twi.h"
 
 static void at91_dbgu_hw_init(void)
 {
@@ -318,6 +319,13 @@ unsigned int at91_twi2_hw_init(void)
 }
 #endif
 
+#if defined(CONFIG_AUTOCONFIG_TWI_BUS)
+void at91_board_config_twi_bus(void)
+{
+	act8865_twi_bus	= 1;
+}
+#endif
+
 #if defined(CONFIG_DISABLE_ACT8865_I2C)
 int at91_board_act8865_set_reg_voltage(void)
 {
diff --git a/board/sama5d4_xplained/sama5d4_xplained.c b/board/sama5d4_xplained/sama5d4_xplained.c
index 5b3cb5c..e530a93 100644
--- a/board/sama5d4_xplained/sama5d4_xplained.c
+++ b/board/sama5d4_xplained/sama5d4_xplained.c
@@ -48,6 +48,7 @@
 #include "tz_utils.h"
 #include "matrix.h"
 #include "act8865.h"
+#include "twi.h"
 
 #if defined(CONFIG_REDIRECT_ALL_INTS_AIC)
 static void redirect_interrupts_to_aic(void)
@@ -509,6 +510,16 @@ unsigned int at91_twi3_hw_init(void)
 }
 #endif
 
+#if defined(CONFIG_AUTOCONFIG_TWI_BUS)
+void at91_board_config_twi_bus(void)
+{
+	hdmi_twi_bus = 0;
+	at24xx_twi_bus = 0;
+
+	act8865_twi_bus = 3;
+}
+#endif
+
 #if defined(CONFIG_DISABLE_ACT8865_I2C)
 int at91_board_act8865_set_reg_voltage(void)
 {
diff --git a/board/sama5d4ek/sama5d4ek.c b/board/sama5d4ek/sama5d4ek.c
index 40e627c..dd1f7d4 100644
--- a/board/sama5d4ek/sama5d4ek.c
+++ b/board/sama5d4ek/sama5d4ek.c
@@ -49,6 +49,7 @@
 #include "tz_utils.h"
 #include "matrix.h"
 #include "act8865.h"
+#include "twi.h"
 
 #if defined(CONFIG_REDIRECT_ALL_INTS_AIC)
 static void redirect_interrupts_to_aic(void)
@@ -521,6 +522,15 @@ unsigned int at91_twi3_hw_init(void)
 }
 #endif
 
+#if defined(CONFIG_AUTOCONFIG_TWI_BUS)
+void at91_board_config_twi_bus(void)
+{
+	hdmi_twi_bus	= 0;
+	wm8904_twi_bus	= 0;
+	act8865_twi_bus	= 0;
+}
+#endif
+
 #if defined(CONFIG_DISABLE_ACT8865_I2C)
 int at91_board_act8865_set_reg_voltage(void)
 {
diff --git a/driver/act8865.c b/driver/act8865.c
index f51a8a7..2ed9860 100644
--- a/driver/act8865.c
+++ b/driver/act8865.c
@@ -41,15 +41,19 @@ static unsigned int act8865_get_twi_bus(void)
 {
 	unsigned int bus = 0;
 
+	if (act8865_twi_bus != 0xff) {
+		bus = act8865_twi_bus;
+	} else {
 #if defined(CONFIG_PMIC_ON_TWI0)
-	bus = 0;
+		bus = 0;
 #elif defined(CONFIG_PMIC_ON_TWI1)
-	bus = 1;
+		bus = 1;
 #elif defined(CONFIG_PMIC_ON_TWI2)
-	bus = 2;
+		bus = 2;
 #elif defined(CONFIG_PMIC_ON_TWI3)
-	bus = 3;
+		bus = 3;
 #endif
+	}
 
 	return bus;
 }
diff --git a/driver/at24xx.c b/driver/at24xx.c
index e0373b6..80df969 100644
--- a/driver/at24xx.c
+++ b/driver/at24xx.c
@@ -38,15 +38,19 @@ static unsigned int at24_get_twi_bus(void)
 {
 	unsigned int bus = 0;
 
+	if (at24xx_twi_bus != 0xff) {
+		bus = at24xx_twi_bus;
+	} else {
 #if defined(CONFIG_EEPROM_ON_TWI0)
-	bus = 0;
+		bus = 0;
 #elif defined(CONFIG_EEPROM_ON_TWI1)
-	bus = 1;
+		bus = 1;
 #elif defined(CONFIG_EEPROM_ON_TWI2)
-	bus = 2;
+		bus = 2;
 #elif defined(CONFIG_EEPROM_ON_TWI3)
-	bus = 3;
+		bus = 3;
 #endif
+	}
 
 	return bus;
 }
diff --git a/driver/at91_twi.c b/driver/at91_twi.c
index 3476641..2389e19 100644
--- a/driver/at91_twi.c
+++ b/driver/at91_twi.c
@@ -35,6 +35,11 @@
 
 unsigned int twi_init_done;
 
+unsigned char hdmi_twi_bus;
+unsigned char wm8904_twi_bus;
+unsigned char act8865_twi_bus;
+unsigned char at24xx_twi_bus;
+
 #if defined(CONFIG_TWI0)
 static unsigned int at91_twi0_base;
 #endif
@@ -296,5 +301,16 @@ void twi_init(void)
 		twi_configure_master_mode(3, bus_clock, TWI_CLOCK);
 #endif
 
+	hdmi_twi_bus	= 0xff;
+	wm8904_twi_bus	= 0xff;
+	act8865_twi_bus	= 0xff;
+	at24xx_twi_bus	= 0xff;
+
+#if defined(CONFIG_AUTOCONFIG_TWI_BUS)
+	dbg_info("Auto-Config the TWI Bus by the board\n");
+
+	at91_board_config_twi_bus();
+#endif
+
 	twi_init_done = 1;
 }
diff --git a/driver/driver_cpp.mk b/driver/driver_cpp.mk
index 7cbbe79..34271b4 100644
--- a/driver/driver_cpp.mk
+++ b/driver/driver_cpp.mk
@@ -316,3 +316,6 @@ endif
 ifeq ($(CONFIG_PM_PMIC), y)
 CPPFLAGS += -DCONFIG_PM_PMIC
 endif
+ifeq ($(CONFIG_AUTOCONFIG_TWI_BUS), y)
+CPPFLAGS += -DCONFIG_AUTOCONFIG_TWI_BUS
+endif
diff --git a/driver/hdmi_SiI9022.c b/driver/hdmi_SiI9022.c
index 9a57abe..557b6ad 100644
--- a/driver/hdmi_SiI9022.c
+++ b/driver/hdmi_SiI9022.c
@@ -62,15 +62,19 @@ static unsigned int SiI9022_get_twi_bus(void)
 {
 	unsigned int bus = 0;
 
+	if (hdmi_twi_bus != 0xff) {
+		bus = hdmi_twi_bus;
+	} else {
 #if defined(CONFIG_HDMI_ON_TWI0)
-	bus = 0;
+		bus = 0;
 #elif defined(CONFIG_HDMI_ON_TWI1)
-	bus = 1;
+		bus = 1;
 #elif defined(CONFIG_HDMI_ON_TWI2)
-	bus = 2;
+		bus = 2;
 #elif defined(CONFIG_HDMI_ON_TWI3)
-	bus = 3;
+		bus = 3;
 #endif
+	}
 
 	return bus;
 }
diff --git a/driver/wm8904.c b/driver/wm8904.c
index 7d139fb..720fe52 100644
--- a/driver/wm8904.c
+++ b/driver/wm8904.c
@@ -62,15 +62,19 @@ static unsigned int wm8904_get_twi_bus(void)
 {
 	unsigned int bus = 0;
 
+	if (wm8904_twi_bus != 0xff) {
+		bus = wm8904_twi_bus;
+	} else {
 #if defined(CONFIG_CODEC_ON_TWI0)
-	bus = 0;
+		bus = 0;
 #elif defined(CONFIG_CODEC_ON_TWI1)
-	bus = 1;
+		bus = 1;
 #elif defined(CONFIG_CODEC_ON_TWI2)
-	bus = 2;
+		bus = 2;
 #elif defined(CONFIG_CODEC_ON_TWI3)
-	bus = 3;
+		bus = 3;
 #endif
+	}
 
 	return bus;
 }
diff --git a/include/board.h b/include/board.h
index 52698e0..64d8f5c 100644
--- a/include/board.h
+++ b/include/board.h
@@ -102,6 +102,7 @@ extern unsigned int at91_twi0_hw_init(void);
 extern unsigned int at91_twi1_hw_init(void);
 extern unsigned int at91_twi2_hw_init(void);
 extern unsigned int at91_twi3_hw_init(void);
+extern void at91_board_config_twi_bus(void);
 
 extern unsigned int at91_eth0_hw_init(void);
 extern unsigned int at91_eth1_hw_init(void);
diff --git a/include/twi.h b/include/twi.h
index 9bf14c0..bd103a8 100644
--- a/include/twi.h
+++ b/include/twi.h
@@ -28,6 +28,11 @@
 #ifndef __TWI_H__
 #define __TWI_H__
 
+extern unsigned char hdmi_twi_bus;
+extern unsigned char wm8904_twi_bus;
+extern unsigned char act8865_twi_bus;
+extern unsigned char at24xx_twi_bus;
+
 extern unsigned int twi_init_done;
 
 extern int twi_read(unsigned int twi_no, unsigned char device_addr,
-- 
1.9.1

